lib/galaxy/model/dataset_collections/builder.py

```diff
@@ -35,23 +75,43 @@ def set_collection_elements(dataset_collection, type, dataset_instances, associa
     return dataset_collection
 
 
+def guess_fields(dataset_instances: Dict[str, Any]) -> List[FieldDict]:
+    fields: List[FieldDict] = []
+    for identifier, element in dataset_instances.items():
+        # TODO: Make generic enough to handle nested record types.
+        assert element.history_content_type == "dataset"
+        fields.append({"type": "File", "name": identifier})
+
```

Comment:
From my latest iteration of https://github.com/galaxyproject/galaxy/pull/12909 :

```suggestion
        if isinstance(element, DatasetCollection):
            return []
        else:
            fields.append({"type": "File", "name": identifier})
```